scripts/bundlebuilder.py

```diff
@@ -41,6 +41,60 @@ def execute(cmd, raise_exception=True):
             return p.returncode, output
 
 
+class Fetcher(object):
+    """
+    Fetcher supports different protocols/means to reach the bundle source.
+    """
+
+    @staticmethod
+    def fetch(repo, branch=None):
+        """
+        Based on the protocol described at the beginning of repo calls
+        the right fetch_* method.
+        Args:
+            repo: repository where the bundle resides
+            branch: if the repo is a git repo, this is the branch
+
+        Returns: a directory where the bundle is
+
+        """
+        if repo.startswith('local:'):
+            path = repo[len('local:'):]
+            return Fetcher.fetch_local(path)
+        else:
+            return Fetcher.fetch_git(repo, branch)
+
+    @staticmethod
+    def fetch_local(path):
+        """
+        Just copy the bundle to a temp location
+        Args:
+            path: the path to the bundle
+
+        Returns: the temp directory containing the bundle
+
+        """
+        tempdir = mkdtemp()
+        destination = "{}/bundle".format(tempdir)
+        copytree(path, destination)
+        return destination
+
+    @staticmethod
+    def fetch_git(repo, branch):
+        """
+        Clone the repository of the bundle
+        Args:
+            repo: the git repo repo with the bundle
+            branch: the branch with the bundle
+
+        Returns: the temp directory containing the bundle
+
+        """
+        tempdir = mkdtemp()
+        execute(["git", "clone", repo, "--branch", branch, "{}/".format(tempdir)])
```

Comment:
Can we also include `--single-branch` as a git param?  This will save a little bit of time by only fetching the git history/metadata for the branch we care about.  As an example, `--single-branch` saves us 18M on the bigtop repo:

```bash
ubuntu@53f55a8f3589:/tmp$ git clone https://github.com/apache/bigtop.git --branch master ./all
Cloning into './all'...
remote: Counting objects: 32936, done.
remote: Compressing objects: 100% (31/31), done.
remote: Total 32936 (delta 13), reused 1 (delta 1), pack-reused 32900
Receiving objects: 100% (32936/32936), 41.50 MiB | 4.84 MiB/s, done.
Resolving deltas: 100% (16289/16289), done.
Checking connectivity... done.

ubuntu@53f55a8f3589:/tmp$ git clone https://github.com/apache/bigtop.git --branch master --single-branch ./single
Cloning into './single'...
remote: Counting objects: 26360, done.
remote: Compressing objects: 100% (33/33), done.
remote: Total 26360 (delta 13), reused 3 (delta 1), pack-reused 26322
Receiving objects: 100% (26360/26360), 23.41 MiB | 4.44 MiB/s, done.
Resolving deltas: 100% (12700/12700), done.
Checking connectivity... done.

ubuntu@53f55a8f3589:/tmp$ du -sh all single
99M	all
81M	single
```